src/main/java/frc/robot/utils/RingBuffer.java

```diff
@@ -0,0 +1,123 @@
+package frc.robot.utils;
+
+import java.util.NoSuchElementException;
+
+public class RingBuffer<T> {
+
+  private Object[] backingArray;
+  private int firstIdx = 0; // Index of the first elements
+  private int nextIdx = 0; // Index of the slot into which we would insert an element on addLast
+  // If nextIdx == (firstIdx - 1) % array_size, then insertion requires resizing the array.
+  // If nextIdx == firstIdx, then the buffer is empty.
+
+  public RingBuffer() {
```

Comment:
rings aren't the game object this year, so the naming of this class needs to be changed at a minimum